harness/determined/core/_checkpoint.py

```diff
@@ -241,79 +243,116 @@ def upload(
                     "cannot call .upload(ckpt_dir=None, shard=False), which would result in doing "
                     "nothing at all"
                 )
-            return self._upload_single(ckpt_dir, metadata)
+            return self._upload_single(ckpt_dir, metadata, selector=selector)
         else:
             storage_id = None
             if self._dist.rank == 0:
                 storage_id = str(uuid.uuid4())
             storage_id = self._dist.broadcast(storage_id)
 
             assert storage_id
-            return self._upload_sharded(ckpt_dir, storage_id, metadata)
+            return self._upload_sharded(ckpt_dir, storage_id, metadata, selector=selector)
 
-    def _upload_single(self, ckpt_dir: str, metadata: Optional[Dict[str, Any]] = None) -> str:
+    def _upload_single(
+        self,
+        ckpt_dir: str,
+        metadata: Optional[Dict[str, Any]] = None,
+        *,
+        selector: Optional[Callable[[str], bool]] = None,
+    ) -> str:
         storage_id = str(uuid.uuid4())
         resources = self._storage_manager._list_directory(ckpt_dir)
 
         # Add metadata pre-upload but without counting it among resources.
         self._write_metadata_file(ckpt_dir, metadata or {})
 
-        self._storage_manager.upload(src=ckpt_dir, dst=storage_id)
+        paths = None
+        if selector is not None:
+            resources = {key: resources[key] for key in resources if selector(key)}
+            paths = set(resources.keys())
```

Comment:
nit: I'm pretty sure `set(resources)` will suffice, the `.keys()` is unnecessary.

Reply:
TIL :)